test/jest/Bladeburner/Actions.test.ts

```diff
@@ -0,0 +1,358 @@
+        expect(Player.money).toBeGreaterThan(before);
+      });
+    });
+
+    /** Stat EXP check for all actions */
+    /** Checking all of them to avoid regressions */
+    describe.each([...actionIdWithIndividualStat(nonGeneralActions())])("$id.name", ({ id, stat }) => {
+      it(`awards ${stat} exp`, () => {
+        guaranteeSuccess(), (before = Player.exp[stat]), complete(id);
+        expect(Player.exp[stat]).toBeGreaterThan(before);
+      });
+    });
+  });
+
+  describe("Upon failed completion", () => {
+    let before;
+
+    describe.each([SampleOperation, SampleBlackOp])("operations and black operations decrease rank", (id) => {
+      it(`${id.type}`, () => {
+        guaranteeFailure(), (before = inst.rank), complete(id);
+        expect(inst.rank).toBeLessThan(before);
+      });
+    });
+  });
+
+  it("have a minimum duration of 1 second", () => {
+    complete(SampleContract);
+    expect(inst.actionTimeToComplete).toBeGreaterThanOrEqual(1);
+  });
+
+  beforeAll(() => {
+    /* Initialise Formatters. Dependency of Bladeburner Logs/Console */
+    FormatsNeedToChange.emit();
+  });
+
+  beforeEach(() => {
+    setPlayer(new PlayerObject());
+
+    /** Need BN5 to receive Int EXP */
+    Player.sourceFiles.set(5, 3);
+
+    if (initBladeburner(Player)) {
+      inst = Player.bladeburner;
+      inst.clearConsole();
+    }
+  });
+
+  function initBladeburner(player: PlayerObject): player is PlayerObject & { bladeburner: Bladeburner } {
+    player.startBladeburner();
+    return true;
+  }
+
+  function basicStats() {
+    inst.rank = 1;
+    inst.changeRank(Player, 10);
+    Player.gainStrengthExp(LOW_STAT_EXP);
+    Player.gainDefenseExp(LOW_STAT_EXP);
+    Player.gainAgilityExp(LOW_STAT_EXP);
+    Player.gainDexterityExp(LOW_STAT_EXP);
+
+    inst.calculateMaxStamina();
+    inst.stamina = inst.maxStamina;
+
+    resetCity();
+  }
+
+  function guaranteeSuccess() {
+    inst.rank = 1;
+    inst.changeRank(Player, 1e10);
+
+    Player.gainStrengthExp(HIGH_STAT_EXP);
+    Player.gainDefenseExp(HIGH_STAT_EXP);
+    Player.gainAgilityExp(HIGH_STAT_EXP);
+    Player.gainDexterityExp(HIGH_STAT_EXP);
+    inst.setSkillLevel(BladeburnerSkillName.BladesIntuition, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.Cloak, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.DigitalObserver, 1e12);
+    inst.setSkillLevel(BladeburnerSkillName.Reaper, 1e12);
+
+    inst.calculateMaxStamina();
+    inst.stamina = inst.maxStamina;
+
+    resetCity();
+  }
+
+  function guaranteeFailure() {
+    inst.rank = 1;
+    inst.changeRank(Player, 400e3); // Minimum to attempt to fail hardest op
+    Player.gainAgilityExp(1e50);
+    inst.calculateMaxStamina();
+    inst.stamina = 0;
+
+    resetCity();
+  }
+
+  function resetCity() {
+    inst.cities[inst.city].chaos = 0;
+    inst.cities[inst.city].comms = 100;
+    inst.cities[inst.city].pop = 1e9;
+
+    /** Disable random event */
+    inst.randomEventCounter = Infinity;
```

Comment:
Random events trigger when the counter is `<= zero`
Don't want random events to interfere with checks for chaos, communities, pops or estimation figures